control-plane/p4RuntimeSerializer.cpp

```diff
@@ -1219,11 +1223,53 @@ class P4RuntimeAnalyzer {
         addAnnotations(profile->mutable_preamble(), actionProfile.annotations);
     }
 
+    void addValueSet(const IR::Declaration_Variable* inst) {
+        // guaranteed by caller
+        CHECK_NULL(inst);
+        BUG_CHECK(inst->type->is<IR::Type_ValueSet>(),
+                  "variable %1% is not of type value_set", inst);
+
+        auto pvsType = inst->type->to<IR::Type_ValueSet>();
+        auto bitwidth = static_cast<uint32_t>(pvsType->width_bits());
+
+        auto name = inst->controlPlaneName();
+
+        unsigned int size = 0;
+        auto sizeAnnotation = inst->getAnnotation("size");
+        if (sizeAnnotation) {
+            if (sizeAnnotation->expr.size() != 1) {
+                ::error("@size should be an integer for declaration %1%", inst);
+                return;
+            }
+            auto sizeConstant = sizeAnnotation->expr[0]->to<IR::Constant>();
+            if (sizeConstant == nullptr || !sizeConstant->fitsInt()) {
+                ::error("@size should be an integer for declaration %1%", inst);
+                return;
+            }
+            if (sizeConstant->value < 0) {
+                ::error("@size should be a positive integer for declaration %1%", inst);
+                return;
+            }
+            size = sizeConstant->value.get_ui();
+        }
+
+        auto vs = p4Info->add_value_sets();
+        auto id = symbols.getId(P4RuntimeSymbolType::VALUE_SET, name);
+        vs->mutable_preamble()->set_id(id);
+        vs->mutable_preamble()->set_name(name);
+        vs->mutable_preamble()->set_alias(symbols.getAlias(name));
+        vs->set_bitwidth(bitwidth);
```

Comment:
Is this the only important thing, or do you care about the structure of the element type?

Reply:
Thanks for the review. At the moment we only care about the total bitwidth but we are reviewing a proposal to include the structure information in P4Info. This is related to the type serialization discussion.